packages/@o3r/extractors/src/core/comparator/metadata-comparator.interface.ts

```diff
@@ -0,0 +1,76 @@
+import type { JsonObject } from '@angular-devkit/core';
+import type { SupportedPackageManagers } from '@o3r/schematics';
+
+/**
+ * Interface of the comparator used to compare 2 different versions of the same metadata file.
+ */
+export interface MetadataComparator<T> {
+  /**
+   * Get an array of metadata items to parse a metadata file content.
+   * @param content Content of a metadata file
+   */
+  getArray(content: any): T[];
+
+  /**
+   * Get a description of a metadata item.
+   * @param item Metadata item
+   */
+  getName(item: T): string;
+
+  /**
+   * Compares 2 metadata items.
+   * @param item1 Metadata item
+   * @param item2 Metadata item to compare with
+   */
+  isSame(item1: T, item2: T): boolean;
+
+  /**
+   * Returns true if a migration item matches a metadata item.
+   * @param metadataItem Metadata item
+   * @param migrationItem Migration item
+   */
+  isMigrationDataMatch(metadataItem: T, migrationItem: any): boolean;
+}
+
+/**
+ * Migration item used to document a migration of a config, localization or styling metadata.
+ */
+export interface MigrationData<T> {
+  /** Metadata type */
+  contentType: 'CONFIG' | 'LOCALIZATION' | 'STYLING';
+
+  /** Previous metadata value */
+  before: T;
+
+  /** New metadata value */
+  after: T;
+}
+
+/**
+ * Generic metadata builder options
+ */
+export interface MigrationMetadataCheckBuilderOptions extends JsonObject {
+  /** Path to the folder containing the migration metadata. */
+  migrationMetadataFolder: string;
+
+  /** Granularity of the migration check. */
+  granularity: 'major' | 'minor';
```

Comment:
`Patch` is not available on purpose?

Reply:
It's not in the requirements but I think I can add it. I will think about it to check that there would be no issue with that

Reply:
I don't think it's useful, it will checks by default that there is no issue with that since you can't merge changes on a released branch (no patch updating the metadata will be allowed)